aten/src/ATen/native/ForeachUtils.h

```diff
@@ -64,155 +64,69 @@ void check_foreach_api_restrictions(TensorList tensors1, TensorList tensors2, Te
 // - All tensors must be non-overlapping and dense
 // - Resulting tensor must have the same dtype as the input one
 
-// Check if all tensors have the same device, layout, strides and are not overlapping and dense
-bool has_same_attributes(Device expected_device, TensorList tensors) {
-  auto expected_strides = tensors[0].strides();
-  for (const auto& t : tensors) {
-    if (t.device() != expected_device) {
-      return false;
-    }
-
-    if (t.layout() != at::kStrided) {
-      return false;
-    }
-
-    if (!t.is_non_overlapping_and_dense()) {
-      return false;
-    }
-
-    if (t.strides() != expected_strides) {
-      return false;
-    }
-  }
-
-  return true;
-}
-
 bool will_promote_tensor(const Tensor& tensor, Scalar scalar) {
   auto result_dtype = at::result_type(tensor, scalar);
   return result_dtype != tensor.scalar_type();
 }
 
-bool can_use_fast_route(TensorList tensors) {
-#ifdef __HIP_PLATFORM_HCC__
-  return false;
-#else
-  auto expected_device = tensors[0].device();
-  for (auto t : tensors) {
-    if (!has_same_attributes(expected_device, {t})) {
-      return false;
-    }
-  }
-
-  return true;
-#endif
-}
-
-bool can_use_fast_route(TensorList tensors, Scalar scalar) {
-#ifdef __HIP_PLATFORM_HCC__
-  return false;
-#else
-  auto expected_device = tensors[0].device();
-
-  for (auto t : tensors) {
-    if (!has_same_attributes(expected_device, {t})) {
-      return false;
+bool check_fast_path_restrictions(
+        ArrayRef<TensorList> tensorLists, 
+        ArrayRef<Scalar> scalarList = {}) {
+    auto expected_device = tensorLists[0][0].device();
+    auto expected_strides = tensorLists[0][0].strides();
+
+    auto is_tensor_okay = [&](const Tensor& tensor) {
+      return tensor.device() == expected_device &&
+             tensor.layout() == at::kStrided &&
+             tensor.strides() == expected_strides &&
+             tensor.is_non_overlapping_and_dense();
+    };
+
+    for (const auto& tensorList : tensorLists) {
+      for (const auto& tensor : tensorList) {
+        if (!is_tensor_okay(tensor)) {
+          return false;
+        }
+      }
     }
 
-    if (will_promote_tensor(t, scalar)) {
-      return false;
+    for (int j=0; j < tensorLists.size(); j++) {
+      for (int i=0; i < tensorLists[j].size(); i++) {
+        if (scalarList.size() == 1) {
+          if (will_promote_tensor(tensorLists[j][i], scalarList[0])) {
+            return false;
+          }
+        } else if (scalarList.size() > 1) {
+          // Complex scalar list is not supported.
+          if (scalarList[i].isComplex()) {
+            return false;
+          }
```

Comment:
Is `_foreach_add([torch.tensor(1.+0j), torch.tensor(-1.+0j)], 0+1j)` supposed to go through the fast path? Why does the fast-path support a single complex scalar but not complex scalarList?

Reply:
TensorListMetadata has to be < 4KB - the limit for kernel launch argument and scalar list with complex scalars will not fit in. So for now, we are pushing it via a slow path. Added a comment about this

Reply:
Got it, thanks for the context! A comment would be great :)